app/components/markdown.tsx

```diff
@@ -116,9 +116,27 @@ function escapeDollarNumber(text: string) {
   return escapedText;
 }
 
+function escapeBrackets(text: string) {
+  const pattern =
+    /(```[\s\S]*?```|`.*?`)|\\\[([\s\S]*?[^\\])\\\]|\\\((.*?)\\\)/g;
```

Comment:
the execution performance of the the block of code is a little bad

Reply:
> the execution performance of the the block of code is a little bad

It's no wonder that regular expressions (regex) often have poor performance, especially for interpreting languages, unlike regex in compiled languages (e.g., regex in Golang, which performs better as it is compiled).

Reply:
At first I was also worried about the performance, but later I found that this worry was unnecessary.
I used the following code to log the time consumed for each call, and the results showed that the function is fast enough.

```javascript
function escapeBrackets(text: string) {
  let begin_time = performance.now();
  const pattern =
    /(```[\s\S]*?```|`.*?`)|\\\[([\s\S]*?[^\\])\\\]|\\\((.*?)\\\)/g;
  let res = text.replace(
    pattern,
    (match, codeBlock, squareBracket, roundBracket) => {
      if (codeBlock) {
        return codeBlock;
      } else if (squareBracket) {
        return `$$${squareBracket}$$`;
      } else if (roundBracket) {
        return `$${roundBracket}$`;
      }
      return match;
    },
  );
  let endTime = performance.now();
  console.log(`escapeBrackets, string length=${text.length}, time consumed=${endTime - begin_time} ms`);
  return res;
}
```
![微信截图_20240326163806](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/28617777/2245f8b2-46d1-4792-aa56-35bf4513744f)
![线性回归模型和评估指标](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/28617777/b10310e5-1c5f-4ad8-9793-9dc2a314da97)

Reply:
Modern js engines compile and cache regexp at load time, so this function will not recompile it every time

Reply:
> At first I was also worried about the performance, but later I found that this worry was unnecessary. I used the following code to log the time consumed for each call, and the results showed that the function is fast enough.
> 
> ```js
> function escapeBrackets(text: string) {
>   let begin_time = performance.now();
>   const pattern =
>     /(```[\s\S]*?```|`.*?`)|\\\[([\s\S]*?[^\\])\\\]|\\\((.*?)\\\)/g;
>   let res = text.replace(
>     pattern,
>     (match, codeBlock, squareBracket, roundBracket) => {
>       if (codeBlock) {
>         return codeBlock;
>       } else if (squareBracket) {
>         return `$$${squareBracket}$$`;
>       } else if (roundBracket) {
>         return `$${roundBracket}$`;
>       }
>       return match;
>     },
>   );
>   let endTime = performance.now();
>   console.log(`escapeBrackets, string length=${text.length}, time consumed=${endTime - begin_time} ms`);
>   return res;
> }
> ```
> 
> ![微信截图_20240326163806](https://private-user-images.githubusercontent.com/28617777/316787839-2245f8b2-46d1-4792-aa56-35bf4513744f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0NjYxMzMsIm5iZiI6MTcxMTQ2NTgzMywicGF0aCI6Ii8yODYxNzc3Ny8zMTY3ODc4MzktMjI0NWY4YjItNDZkMS00NzkyLWFhNTYtMzViZjQ1MTM3NDRmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI2VDE1MTAzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM5ODZiYmYxM2JjZDJlOWE4MWYxMzE0NjJmMTNmNTM4OGRhZWJjNzUzZmRjODdlZjNhN2I1OTUzOTkyZDFhMTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cB8OzrKyOTeFogM7u_W01_PxyX8J6D-dMy8PqeXa9Qs) ![线性回归模型和评估指标](https://private-user-images.githubusercontent.com/28617777/316787872-b10310e5-1c5f-4ad8-9793-9dc2a314da97.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0NjYxMzMsIm5iZiI6MTcxMTQ2NTgzMywicGF0aCI6Ii8yODYxNzc3Ny8zMTY3ODc4NzItYjEwMzEwZTUtMWM1Zi00YWQ4LTk3OTMtOWRjMmEzMTRkYTk3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI2VDE1MTAzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZkNTAxODM4NzBlYWNlZmZmY2M3YjBiYjgxYzIzNzBlMzAzM2UxZDk1NmIyYmQwYjQ3ZDhmZGUwZGVlYTQ2NmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.THrj-vPiztBb9WXADTTdDms236-YFFPZojZhyxGA160)

it because got handle by react `useMemo` that's why it looks faster

![image](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/17626300/8063fa26-c203-45de-aa08-4c315eef8bef)